proposals/UIWebView-strategy.md

```diff
@@ -0,0 +1,86 @@
+UIWebView Strategy for Cordova iOS
+==================================
+
+As of August 2019, Apple is now showing a deprecation warning when uploading apps to the App Store that include UIWebView-related code. UIWebView has been unofficially deprecated in favour of the new WKWebView for a while, but Apple has now made it officially deprecated. As a result, all Cordova apps built for iOS will receive this deprecation warning on upload. Currently it is just a warning, but Apple has not provided a timeline for when it will start rejecting apps that use UIWebView APIs.
+
+We've had [a plan](https://cordova.apache.org/news/2018/08/01/future-cordova-ios-webview.html) for several years to make it easier to migrate to WKWebView and eventually deprecate UIWebView use by Cordova, but almost no progress has been made on this objective since the plan was outlined over a year ago.
+
+Cordova users are clamoring for a fix to the problem and there are pull requests open to remove UIWebView entirely, but if we proceed without a concrete plan then we are guaranteed to cause a lot of pain and suffering for Cordova users.
+
+The goal of this document is to outline several potential migration plans, and facilitate discussion on the pros and cons of each.
+
+
+High-Level Goals
+----------------
+
+* Make WKWebView the default web view for Cordova iOS applications
+* Make it possible to build Cordova iOS apps without including any UIWebView APIs
+* Maintain compatibility as much as possible with existing plugins
+* Provide a means for Cordova apps to transition from UIWebView to WKWebView
+
+
+Challenges
+----------
+
+* UIWebView-related classes are potentially part of the public API exposed to plugins
+* Moving all apps over to WKWebView will result in data being lost (localStorage, indexedDB, etc.)
+* WKWebView does not support all the same features as UIWebView and has stricter security requirements, and **many existing Cordova iOS apps will not work if run in WKWebView** without changes
+* The current WKWebView plugin needs updates to use Scheme Handlers to avoid issues that require a local web server.
+* The InAppBrowser plugin also uses UIWebView. Moving to WKWebView would potentially lose some functionality.
+    * Ideally, the InAppBrowser should be deprecated in favour of SFSafariViewController which severely restricts functionality in the name of security. Currently InAppBrowser use could largely be considered bad practice.
```

Comment:
Very much different use case and functionality, also not related to UIWebView<->WKWebView. This should not be part of this discussion.

What probably should be part, is how the default InAppBrowser would break when UIWebView is removed from `cordova-ios`.

What should also be part of the discussion, is that IAP should probably do the same for dropping UIWebView and switch to WKWebView as default.

Reply:
> Very much different use case and functionality, also not related to UIWebView<->WKWebView. This should not be part of this discussion.

Agreed, but I also feel that deprecating IAB is in line with "the goal of [Cordova] is to cease to exist" now that there are platform-supported better options.
I'll drop this point from the discussion.

Reply:
It's definitely a better option for what IAB was originally created, safe web browsing.
But as security is its strong suit, it's not as feature rich as IAB, all code/css injection, events, etc. that people relies on nowadays is not possible, so we can't deprecate it.

Reply:
I agree. I have to maintain some pages which rely on that for some time. If you want your pages in IAB tightly integrated to your app you need this script injection stuff. Deprecating it should be the last resort, since it works fine now if you know what you are doing and use it in a responsible way. So security is not really a point.

Reply:
the 'code/css injection' was always controversial, I was always against it as I wanted it to be a separate sandbox.
Pushing this off to wkwv is probably fine, and making a firm recommendation of SFSafariViewController when apps just need static pages, and/or oauth.